### PR TITLE
Prototype of a module with Firestore fakes for testing dependent code in Swift

### PR DESCRIPTION
**Pull request by maksymmalyhin**
An alternative approach to #7660 to enable testing code dependent on Firestore in Swift.

### CODE REVIEW

File: FirebaseFirestoreTestingSupport.podspec
Comment (peterfriese) [inline FirebaseFirestoreTestingSupport.podspec:2]: Should we
a) define a pod for each of the products we support (i.e. FirebaseFirestoreTestingSupport, FirebaseAuthTestingSupport, etc.)
b) define a single pod (FirebaseTestingSupport)

File: Package.swift
Comment (peterfriese) [inline Package.swift:844]: Similar to the question about a single vs multiple pods, should we define a single vs. multiple Swift packages?

File: FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift
Comment (peterfriese) [inline FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift:18]: I assume we would put the tests inside FirebaseCombineSwift/Tests

I've tested this, and it works (of course, we need to set a dependency from the FirebaseCombineSwift pod to the testing support pod),

**Review by peterfriese — COMMENTED** [review]
This looks great, I like this a lot better than all the other approaches we've tried so far!

File: FirebaseFirestoreTestingSupport.podspec
Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec:2]: Good question! Initially I wanted to have a single FirebaseTestingSupport spec but discovered the following disadvantages:
- with a single podspec we will have to add unnecessary dependencies to the test targets (e.g. we probably don't want adding Firestore dependency to Auth tests)
- OS versions requirements may be different for different SDKs. It's harder to express with a single FirebaseTestingSupport spec.

So my proposal is to have separate podspecs/SPM targets for testing support code per SDK.

WDYT?

File: Package.swift
Comment (maksymmalyhin) [reply Package.swift:844]: I think the [response](https://github.com/firebase/firebase-ios-sdk/pull/7668#discussion_r589634538) is relevant here - I think separate targets are better.

File: FirebaseFirestoreTestingSupport.podspec
Comment (peterfriese) [reply FirebaseFirestoreTestingSupport.podspec:2]: SG - happy with the extra flexibility this gives us.

File: FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift
Comment (maksymmalyhin) [reply FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift:18]: This test is for fake implementation only and may be not 100% required, but it may help in the future debug the test failures and understand if it is the fake implementation issue or its client issue.

Comment (maksymmalyhin) [reply FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift:18]: Yeah, the testing support targets should be added as dependencies to the test targets which is not demonstrated in the current version of the PR.

Comment (maksymmalyhin) [reply FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift:18]: Good point on the docs! Will eventually add them, but feel free to start using it if it unblocks the Combine effort.

File: FirebaseFirestoreTestingSupport.podspec
Comment (paulb777) [inline FirebaseFirestoreTestingSupport.podspec:1]: If the podspec is only for testing and not for publishing, it might be better to organize in a lower level directory like how Firestore manages LibFuzzer and GoogleBenchmark in Firestore/Example.

File: FirebaseFirestoreTestingSupport.podspec
Comment (paulb777) [inline FirebaseFirestoreTestingSupport.podspec:3]: How about 1.0.0.

I'm not sure we should tie the version to Firebase versioning?

**Review by paulb777 — COMMENTED** [review]
Let's sort out the directory structure and then LGTM

File: FirebaseFirestoreTestingSupport.podspec
Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec:3]: Good idea, will do.

File: FirebaseFirestoreTestingSupport.podspec
Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec:1]: I like the idea! Let me try to restructure it to see if it works well.

File: FirebaseFirestoreTestingSupport.podspec
Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec]: Moved under FirebaseTestingSupport folder.

Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec]: Hm, just realized that having the podspec in a subfolder will require `scripts/pod_lib_lint.rb` modifications to accommodate it. I'll look into it but we may want to move the podspec back to the root dir temporary to unblock testing.

Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec]: I'm moving the podspec back to the root dir so far as we do some assumptions in `scripts/pod_lib_lint.rb` that make subfolder support not trivial (not impossible, but just requires a bit of refactoring). 

Another thought, as we may offer the testing support to developer, we may want to keep the spec in the root dir. Let's discuss it a bit more and apply changes if needed in the following PRs.

File: FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift
Comment (maksymmalyhin) [reply FirebaseTestingSupport/Firestore/Tests/QueryFakeTests.swift:18]: I added a README file with brief explanations and common steps, PTAL.

File: FirebaseTestingSupport/README.md
Comment (paulb777) [inline FirebaseTestingSupport/README.md:40]: Extra comma?

File: FirebaseTestingSupport/README.md
Comment (paulb777) [inline FirebaseTestingSupport/README.md:51]: s/targte/target/

File: FirebaseTestingSupport/README.md
Comment (paulb777) [inline FirebaseTestingSupport/README.md:1]: Clarify that these instructions are for test creators not test users.

**Review by paulb777 — APPROVED** [review]
LGTM on comment resolution and green

File: FirebaseTestingSupport/README.md
Comment (maksymmalyhin) [reply FirebaseTestingSupport/README.md:40]: This is the command for the case when the spec is in the subfolder. The docs are becoming outdated even before the PR merging 😄.

File: FirebaseTestingSupport/README.md
Comment (maksymmalyhin) [reply FirebaseTestingSupport/README.md]: Updated, PTAL.

File: FirebaseFirestoreTestingSupport.podspec
Comment (paulb777) [inline FirebaseFirestoreTestingSupport.podspec:40]: I don't think we want preserve_paths since they break Xcode completion and we don't need to worry about tests mapping up an end user's header map (the reason they're there in the Firestore pod)

File: FirebaseFirestoreTestingSupport.podspec
Comment (paulb777) [inline FirebaseFirestoreTestingSupport.podspec:46]: Might as well start with the current 7.7 or 7.8 since we'll never test earlier versions.

File: FirebaseTestingSupport/README.md
Comment (paulb777) [inline FirebaseTestingSupport/README.md:3]: Extra space before "and"

**Review by paulb777 — APPROVED** [review]

File: FirebaseFirestoreTestingSupport.podspec
Comment (maksymmalyhin) [reply FirebaseFirestoreTestingSupport.podspec]: Good catch! It is unnecessary leftovers from the attempts to fix C++ dependent code.

Thank you for the thorough review!

**Comment by maksymmalyhin** [comment]
@peterfriese The PR looks in a good shape. Feel free to merge it or let me know when you would like me to do it.

**Comment by peterfriese** [comment]
Will get to it later today, thanks for putting this together!

**Review by peterfriese — APPROVED** [review]
This looks great, thanks @maksymmalyhin!

**Pull request merged**
